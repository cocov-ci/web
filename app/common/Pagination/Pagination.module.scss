@import "styles/variables";

.pagination {
  width: 100%;
  text-align: center;
  margin-bottom: 60px;
  margin-top: 50px;

  .pages {
    list-style-type: none;
    display: inline-flex;
    align-items: center;

    .arrow {
      cursor: pointer;
      margin: 0 8px;

      &.disabled {
        color: $colorGray;
      }
    }

    .page {
      font-size: 17px;
      border-radius: 8px;
      margin: 0 4px;
      width: 37px;
      height: 37px;
      display: flex;
      align-items: center;
      justify-content: center;
      cursor: pointer;

      &.active {
        border: 1px solid $colorGray;
        font-variation-settings: "wght" 500;
        background-color: $colorGrayLighter;
      }

      &:hover {
        border: 1px solid $colorGray;
      }
    }
  }
}
