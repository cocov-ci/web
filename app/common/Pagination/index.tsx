'use client'

import classNames from 'classnames'
import { ChevronFirst, ChevronLast } from 'lucide-react'
import React from 'react'

import styles from './Pagination.module.scss'

export interface PaginationProps {
  total: number
  currentPage: number
  onPageClick: (page: number) => void
  className?: string
}

interface ArrowProps {
  direction: 'previous' | 'next'
  disabled: boolean
  onClick: () => void
}

const Arrow = ({ direction, disabled, onClick }: ArrowProps) => {
  const Component = direction === 'previous' ? ChevronFirst : ChevronLast

  return (
    <Component
      className={classNames(styles.arrow, {
        [styles.disabled]: disabled,
      })}
      onClick={onClick}
    />
  )
}

const slidingPaginationWindow = (numPages: number, current: number) => {
  const maxPerWindow = 9 // Odd is better!
  let pages: number[]

  if (current > maxPerWindow / 2 && numPages > maxPerWindow) {
    const around = (maxPerWindow - 1) / 2
    const first = Math.max(current - around, 1)
    const last = current + around

    const result = []

    for (let i = first; i <= Math.min(last, numPages); i++) {
      result.push(i)
    }

    if (result.length < maxPerWindow) {
      const upTo = result[0]

      for (let i = upTo - (maxPerWindow - result.length); i < upTo; i++) {
        result.unshift(i)
      }
    }

    pages = result
  } else {
    pages = [...Array(numPages)].map((i, idx) => idx + 1)
  }

  if (pages.length > maxPerWindow) {
    pages = pages.slice(0, maxPerWindow)
  }

  return pages
}

const Pagination = ({
  total,
  currentPage,
  onPageClick,
  className,
}: PaginationProps) => {
  const firstPage = currentPage === 1
  const lastPage = currentPage === total

  return (
    <div className={classNames(styles.pagination, className)}>
      <ul className={styles.pages}>
        <Arrow
          direction="previous"
          disabled={firstPage}
          onClick={() =>
            !firstPage && onPageClick && onPageClick(currentPage - 1)
          }
        />

        {slidingPaginationWindow(total, currentPage).map(
          (page): React.ReactElement => {
            return (
              <li
                className={classNames(styles.page, {
                  [styles.active]: page === currentPage,
                })}
                key={page}
                onClick={() => onPageClick && onPageClick(page)}
              >
                {page}
              </li>
            )
          },
        )}

        <Arrow
          direction="next"
          disabled={lastPage}
          onClick={() =>
            !lastPage && onPageClick && onPageClick(currentPage + 1)
          }
        />
      </ul>
    </div>
  )
}

export default Pagination
