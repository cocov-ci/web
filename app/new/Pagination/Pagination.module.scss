@import "styles/variables.module";

.paging {
  border-top: 1px solid $colorGray;

  .paginationComponent {
    padding: 10px 0;
    margin: 0;
  }
}
