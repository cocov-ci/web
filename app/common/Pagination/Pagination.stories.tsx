import type { Meta, StoryObj } from '@storybook/react'

import Pagination from './index'

const meta: Meta<typeof Pagination> = {
  title: 'Common/Pagination',
  component: Pagination,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof Pagination>

export const Default: Story = {
  args: {
    total: 10,
    currentPage: 1,
  },
}

export const Empty: Story = {
  args: {
    total: 0,
    currentPage: 1,
  },
}

export const LargeBegin: Story = {
  args: {
    total: 44,
    currentPage: 1,
  },
}

export const LargeMiddle: Story = {
  args: {
    total: 44,
    currentPage: 22,
  },
}

export const LargeEnd: Story = {
  args: {
    total: 44,
    currentPage: 44,
  },
}
