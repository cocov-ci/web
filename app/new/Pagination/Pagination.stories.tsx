import type { Meta, StoryObj } from '@storybook/react'

import Pagination from './index'

const meta: Meta<typeof Pagination> = {
  title: 'Repositories/New/Pagination',
  component: Pagination,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof Pagination>

export const Default: Story = {
  args: {},
}
