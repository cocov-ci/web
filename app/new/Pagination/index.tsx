'use client'

import Pagination from 'app/common/Pagination'

import styles from './Pagination.module.scss'

interface ReposPaginationParams {
  total: number
  currentPage: number
  onPageClick: (arg: number) => void
}

const ReposPagination = ({
  total,
  currentPage,
  onPageClick,
}: ReposPaginationParams) => {
  return (
    <div className={styles.paging}>
      <Pagination
        className={styles.paginationComponent}
        currentPage={currentPage}
        onPageClick={onPageClick}
        total={total}
      />
    </div>
  )
}

export default ReposPagination
