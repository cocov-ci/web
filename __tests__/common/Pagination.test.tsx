import { fireEvent, render, screen } from '@testing-library/react'

import Pagination from 'app/common/Pagination'

import '@testing-library/jest-dom'

const args = {
  total: 10,
  currentPage: 2,
  onPageClick: () => null,
}

describe('common/Pagination', () => {
  it('renders Pagination', () => {
    const { container } = render(<Pagination {...args} />)

    expect(container.querySelectorAll('.pages li').length).toEqual(10)
    expect(container.querySelectorAll('svg').length).toEqual(2)
  })

  it('triggers onPageClick event', () => {
    const onPageClick = jest.fn()
    render(<Pagination {...args} onPageClick={onPageClick} />)

    fireEvent.click(screen.getByText('3'))
    expect(onPageClick).toHaveBeenCalledTimes(1)
  })

  it('triggers onPageClick event on Previous arrow', () => {
    const onPageClick = jest.fn()
    const { container } = render(
      <Pagination {...args} onPageClick={onPageClick} />,
    )

    fireEvent.click(container.querySelectorAll('svg')[0] as Element)
    expect(onPageClick).toHaveBeenCalledTimes(1)
  })

  it('triggers onPageClick event on Next arrow', () => {
    const onPageClick = jest.fn()
    const { container } = render(
      <Pagination {...args} onPageClick={onPageClick} />,
    )

    fireEvent.click(container.querySelectorAll('svg')[1] as Element)
    expect(onPageClick).toHaveBeenCalledTimes(1)
  })

  it('renders Pagination snapshots', () => {
    const { container } = render(<Pagination {...args} />)

    expect(container).toMatchSnapshot()
  })
})
