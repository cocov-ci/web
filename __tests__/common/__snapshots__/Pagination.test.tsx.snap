// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Pagination renders Pagination snapshots 1`] = `
<div>
  <div
    class="pagination"
  >
    <ul
      class="pages"
    >
      <svg
        class="arrow"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="17 18 11 12 17 6"
        />
        <path
          d="M7 6v12"
        />
      </svg>
      <li
        class="page"
      >
        1
      </li>
      <li
        class="page active"
      >
        2
      </li>
      <li
        class="page"
      >
        3
      </li>
      <li
        class="page"
      >
        4
      </li>
      <li
        class="page"
      >
        5
      </li>
      <li
        class="page"
      >
        6
      </li>
      <li
        class="page"
      >
        7
      </li>
      <li
        class="page"
      >
        8
      </li>
      <li
        class="page"
      >
        9
      </li>
      <li
        class="page"
      >
        10
      </li>
      <svg
        class="arrow"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="7 18 13 12 7 6"
        />
        <path
          d="M17 6v12"
        />
      </svg>
    </ul>
  </div>
</div>
`;
